client/core/core.go

```diff
@@ -10052,20 +10052,49 @@ func (c *Core) ShieldFunds(assetID uint32, amt uint64) (dex.Bytes, error) {
 
 // UnshieldFunds moves funds from the shielded account to the transparent
 // account.
-func (c *Core) UnshieldFunds(assetID uint32, amt uint64) (dex.Bytes, error) {
+func (c *Core) UnshieldFunds(assetID uint32, amt uint64) ([]byte, error) {
 	sw, err := c.shieldedWallet(assetID)
 	if err != nil {
 		return nil, err
 	}
+
 	return sw.UnshieldFunds(c.ctx, amt)
 }
 
 // SendShielded sends funds from the shielded account to the provided shielded
 // or transparent address.
-func (c *Core) SendShielded(assetID uint32, toAddr string, amt uint64) (dex.Bytes, error) {
+func (c *Core) SendShielded(appPW []byte, assetID uint32, toAddr string, amt uint64) ([]byte, error) {
+	_, err := c.encryptionKey(appPW)
```

Comment:
Ohhhh...  is the use case more like: you walk away from your computer with dex opened and logged in, someone walks up to it and tries to send funds, so a password prompt is there to prevent that?
If so maybe at some point we build a purely frontend mechanism for checking the password based on a hash of the correct pass used on login.

Reply:
> Ohhhh... is the use case more like: you walk away from your computer with dex opened and logged in, someone walks up to it and tries to send funds, so a password prompt is there to prevent that?

Yes.

> If so maybe at some point we build a purely frontend mechanism for checking the password based on a hash of the correct pass used on login.

I guess. Front-end would need to add additional unlock wallet calls in some places, and webserver would need to track password changes obviously. The RPC server would need to mirror all of the changes. It's not a clear win for me, though I understand the desire.

Reply:
I think guarding against a desktop that is not physically secure is fairly unnecessary, but alright.

> Front-end would need to add additional unlock wallet calls in some places, and webserver would need to track password changes obviously.

I'm talking about places where we do this only for the extra auth, _not_ when it's used to unlock the wallet.

This doesn't pertain to the RPC server.  Each RPC is authorized with the rpcuser/pass, and as I was describing above, the only thing Core needs to check is if the user had logged in prior.

Reply:
What if the password was cached in core on login, perhaps optionally with a timeout? This is what dcrwallet does with the `walletpassphrase` command.

Reply:
I'm really not concerned with places where the crypter from the appPW is actually used, but the cases where we're using the backend to add a second auth.  Note the line of code this thread is commenting on.  It's simply a redundant authorization.

This is getting off on a tangent though because we've established the Send methods really might actually want to unlock on the fly, and I'm really fine with having them take a password for that use.  I'm just trying to figure out our policy of when we have a `Core` method check a password if it doesn't go anywhere else after the check.

Reply:
I agree that there's no point to do redundant checks in core.

I don't mean to get too of topic, but the reason I was thinking about the caching was that by moving the bots out of `core`, the bot package needs to cache the password, because it needs to be sent to `Trade`, which unlocks wallets if they have not yet been unlocked. A password is also needed for `Cancel`, but that just does the redundant auth which can be removed.

Reply:
If `Core` would become a go interface that other 3rd party applications could potentially program against, I think it would be a nicer interface without the additional password arguments after login.

Reply:
> What if the password was cached in core on login, perhaps optionally with a timeout? This is what dcrwallet does with the `walletpassphrase` command.

Please no. The RPC auth model is not appropriate for a universal-consumer interface like **core**.

> If `Core` would become a go interface that other 3rd party applications could potentially program against, I think it would be a nicer interface without the additional password arguments after login.

**core** already is a Go interface that other "3rd party apps" can program against. In fact, that's its entire purpose.

Reply:
> **core** already is a Go interface that other "3rd party apps" can program against. In fact, that's its entire purpose

That we all agree on.

The concerns of authorizing certain requests is all that's a bit misplaced sometimes (not talking about combining wallet unlock for convenience, which is generally ok).  Consider: Go consumer X, a middleware like webserver, provides the correct app pass to (*Core).Login.  Later, same Go consumer must provide it again for other things.  These are compile time relations, in the same application and memory space, unlike external connections like rpc/http.  So we just have some minor separation of concerns issues.  I think that's all martop is referring to in terms of the usability of the Core API.

Not to worry though -- I do NOT think caching the app password in Core is particularly relevant or appropriate.  This is either the job of middleware (webserver/rpcserver) or frontend itself.  An external bot can handle it imo without Core assistance.

The only thing I keep coming back to is getting clear about what core methods should even require an app password.  The answer in my options is those that use it, not just check it.